diff --git a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataExportParams.java b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataExportParams.java
index 4081d18a7..388f7100e 100644
--- a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataExportParams.java
+++ b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataExportParams.java
@@ -34,6 +34,7 @@ import com.google.common.collect.Sets;
 
 import org.hisp.dhis.common.IdSchemes;
 import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dataelement.DataElementCategoryOptionCombo;
 import org.hisp.dhis.dataelement.DataElementGroup;
 import org.hisp.dhis.dataset.DataSet;
 import org.hisp.dhis.organisationunit.OrganisationUnit;
@@ -64,6 +65,8 @@ public class DataExportParams
     private boolean includeChildren;
 
     private Set<OrganisationUnitGroup> organisationUnitGroups = new HashSet<>();
+
+    private Set<DataElementCategoryOptionCombo> attributeOptionCombos = new HashSet<>();
     
     private boolean includeDeleted;
 
@@ -107,6 +110,11 @@ public class DataExportParams
         return periods != null && !periods.isEmpty() ? periods.iterator().next() : null;
     }
 
+    public boolean hasPeriods()
+    {
+        return periods != null && !periods.isEmpty();
+    }
+    
     public boolean hasStartEndDate()
     {
         return startDate != null && endDate != null;
@@ -126,12 +134,12 @@ public class DataExportParams
     {
         return organisationUnitGroups != null && !organisationUnitGroups.isEmpty();
     }
-
-    public boolean hasPeriods()
+    
+    public boolean hasAttributeOptionCombos()
     {
-        return periods != null && !periods.isEmpty();
+        return attributeOptionCombos != null && !attributeOptionCombos.isEmpty();
     }
-    
+
     public boolean hasLastUpdated()
     {
         return lastUpdated != null;
@@ -166,6 +174,7 @@ public class DataExportParams
             add( "org units", organisationUnits ).
             add( "children", includeChildren ).
             add( "org unit groups", organisationUnitGroups ).
+            add( "attribute option combos", attributeOptionCombos ).
             add( "deleted", includeDeleted ).
             add( "output id schemes", outputIdSchemes ).toString();
     }
@@ -262,6 +271,17 @@ public class DataExportParams
         return this;
     }
 
+    public Set<DataElementCategoryOptionCombo> getAttributeOptionCombos()
+    {
+        return attributeOptionCombos;
+    }
+
+    public DataExportParams setAttributeOptionCombos( Set<DataElementCategoryOptionCombo> attributeOptionCombos )
+    {
+        this.attributeOptionCombos = attributeOptionCombos;
+        return this;
+    }
+
     public boolean isIncludeDeleted()
     {
         return includeDeleted;
diff --git a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetService.java b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetService.java
index 0626a1ec8..d5a600bef 100644
--- a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetService.java
+++ b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetService.java
@@ -48,8 +48,28 @@ import java.util.Set;
  */
 public interface DataValueSetService
 {
+    /**
+     * Returns a data export object for the given parameters.
+     * 
+     * @param dataSets data sets.
+     * @param dataElementGroups the data element groups.
+     * @param periods the periods.
+     * @param startDate the start date.
+     * @param endDate the end date.
+     * @param organisationUnits the organisation units.
+     * @param includeChildren whether to include org unit children.
+     * @param organisationUnitGroups the organisation unit groupps.
+     * @param attributeOptionCombos the attribute option combos.
+     * @param includeDeleted whether to include deleted data values.
+     * @param lastUpdated filter data values updated after a time stamp.
+     * @param lastUpdatedDuration the last updated duration filter.
+     * @param limit max number of data values to return.
+     * @param idSchemes the identifier schemes.
+     * @return
+     */
     DataExportParams getFromUrl( Set<String> dataSets, Set<String> dataElementGroups, Set<String> periods, Date startDate, Date endDate, 
-        Set<String> organisationUnits, boolean includeChildren, Set<String> organisationUnitGroups, boolean includeDeleted, Date lastUpdated, String lastUpdatedDuration, Integer limit, IdSchemes idSchemes );
+        Set<String> organisationUnits, boolean includeChildren, Set<String> organisationUnitGroups, Set<String> attributeOptionCombos,
+        boolean includeDeleted, Date lastUpdated, String lastUpdatedDuration, Integer limit, IdSchemes idSchemes );
     
     void validate( DataExportParams params );
     
diff --git a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java
index d57b1de8b..00704a675 100644
--- a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java
+++ b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java
@@ -188,8 +188,8 @@ public class DefaultDataValueSetService
 
     @Override
     public DataExportParams getFromUrl( Set<String> dataSets, Set<String> dataElementGroups, Set<String> periods, Date startDate, Date endDate,
-        Set<String> organisationUnits, boolean includeChildren, Set<String> organisationUnitGroups, boolean includeDeleted, Date lastUpdated,
-        String lastUpdatedDuration, Integer limit, IdSchemes outputIdSchemes )
+        Set<String> organisationUnits, boolean includeChildren, Set<String> organisationUnitGroups, Set<String> attributeOptionCombos,
+        boolean includeDeleted, Date lastUpdated, String lastUpdatedDuration, Integer limit, IdSchemes outputIdSchemes )
     {
         DataExportParams params = new DataExportParams();
 
@@ -205,6 +205,8 @@ public class DefaultDataValueSetService
                 DataElementGroup.class, IdentifiableProperty.UID, dataElementGroups ) );
         }
 
+        System.out.println( "PERIODS " + periods );
+        
         if ( periods != null && !periods.isEmpty() )
         {
             params.getPeriods().addAll( periodService.reloadIsoPeriods( new ArrayList<>( periods ) ) );
@@ -216,6 +218,10 @@ public class DefaultDataValueSetService
                 .setEndDate( endDate );
         }
 
+        System.out.println( "RELOADED " + params.getPeriods() );
+        for ( Period period : params.getPeriods())
+            System.out.println( period.getId() );
+        
         if ( organisationUnits != null )
         {
             params.getOrganisationUnits().addAll( identifiableObjectManager.getObjects(
@@ -227,6 +233,12 @@ public class DefaultDataValueSetService
             params.getOrganisationUnitGroups().addAll( identifiableObjectManager.getObjects(
                 OrganisationUnitGroup.class, IdentifiableProperty.UID, organisationUnitGroups ) );
         }
+        
+        if ( attributeOptionCombos != null )
+        {
+            params.getAttributeOptionCombos().addAll( identifiableObjectManager.getObjects( 
+                DataElementCategoryOptionCombo.class, IdentifiableProperty.UID, attributeOptionCombos ) );
+        }
 
         return params
             .setIncludeChildren( includeChildren )
diff --git a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java
index 95ef66763..2ecd7321b 100644
--- a/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java
+++ b/dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java
@@ -298,7 +298,7 @@ public class SpringDataValueSetStore
 
             sql += ") ";
         }
-
+        
         if ( !params.isIncludeDeleted() )
         {
             sql += "and dv.deleted is false ";
@@ -313,6 +313,11 @@ public class SpringDataValueSetStore
             sql += "and dv.periodid in (" + getCommaDelimitedString( getIdentifiers( params.getPeriods() ) ) + ") ";
         }
 
+        if ( params.hasAttributeOptionCombos() )
+        {
+            sql += "and dv.attributeoptioncomboid in (" + getCommaDelimitedString( getIdentifiers( params.getAttributeOptionCombos() ) ) + ") ";
+        }
+
         if ( params.hasLastUpdated() )
         {
             sql += "and dv.lastupdated >= '" + getLongGmtDateString( params.getLastUpdated() ) + "' ";
diff --git a/dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceExportTest.java b/dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceExportTest.java
index 42e6711fb..19d064dab 100644
--- a/dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceExportTest.java
+++ b/dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceExportTest.java
@@ -174,19 +174,19 @@ public class DataValueSetServiceExportTest
         // Data values
 
         dataValueService.addDataValue( new DataValue( deA, peA, ouA, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deA, peA, ouA, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deA, peA, ouA, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
         dataValueService.addDataValue( new DataValue( deA, peA, ouB, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deA, peA, ouB, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deA, peA, ouB, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
 
         dataValueService.addDataValue( new DataValue( deA, peB, ouA, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deA, peB, ouA, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deA, peB, ouA, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
         dataValueService.addDataValue( new DataValue( deA, peB, ouB, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deA, peB, ouB, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deA, peB, ouB, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
 
         dataValueService.addDataValue( new DataValue( deB, peA, ouA, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deB, peA, ouA, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deB, peA, ouA, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
         dataValueService.addDataValue( new DataValue( deB, peA, ouB, cocA, cocA, "1", "storedBy", new Date(), "comment" ) );
-        dataValueService.addDataValue( new DataValue( deB, peA, ouB, cocB, cocA, "1", "storedBy", new Date(), "comment" ) );
+        dataValueService.addDataValue( new DataValue( deB, peA, ouB, cocB, cocB, "1", "storedBy", new Date(), "comment" ) );
 
         // Service mocks
 
@@ -228,6 +228,33 @@ public class DataValueSetServiceExportTest
             assertEquals( peA.getUid(), dv.getPeriod() );
         }
     }
+    
+    @Test
+    public void testExportAttributeOptionCombo()
+    {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        DataExportParams params = new DataExportParams()
+            .setDataSets( Sets.newHashSet( dsA ) )
+            .setOrganisationUnits( Sets.newHashSet( ouB ) )
+            .setPeriods( Sets.newHashSet( peA ) )
+            .setAttributeOptionCombos( Sets.newHashSet( cocA ) );
+
+        dataValueSetService.writeDataValueSetJson( params, out );
+
+        DataValueSet dvs = JacksonUtils.fromJson( out.toByteArray(), DataValueSet.class );
+
+        assertNotNull( dvs );
+        assertNotNull( dvs.getDataSet() );
+        assertEquals( 2, dvs.getDataValues().size() );
+
+        for ( org.hisp.dhis.dxf2.datavalue.DataValue dv : dvs.getDataValues() )
+        {
+            assertNotNull( dv );
+            assertEquals( ouB.getUid(), dv.getOrgUnit() );
+            assertEquals( peA.getUid(), dv.getPeriod() );
+        }        
+    }
 
     @Test
     public void testExportOrgUnitChildren()
@@ -325,7 +352,6 @@ public class DataValueSetServiceExportTest
             assertNotNull( dv );
             assertEquals( avA.getValue(), dv.getDataElement() );
             assertEquals( avB.getValue(), dv.getOrgUnit() );
-            assertEquals( avC.getValue(), dv.getAttributeOptionCombo() );
         }
     }
 }
diff --git a/dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/DataValueSetController.java b/dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/DataValueSetController.java
index 9e1f40de4..3256b5b4e 100644
--- a/dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/DataValueSetController.java
+++ b/dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/DataValueSetController.java
@@ -109,6 +109,7 @@ public class DataValueSetController
         @RequestParam( required = false ) Set<String> orgUnit,
         @RequestParam( required = false ) boolean children,
         @RequestParam( required = false ) Set<String> orgUnitGroup,
+        @RequestParam( required = false ) Set<String> attributeOptionCombo,
         @RequestParam( required = false ) boolean includeDeleted,
         @RequestParam( required = false ) Date lastUpdated,
         @RequestParam( required = false ) String lastUpdatedDuration,
@@ -118,7 +119,8 @@ public class DataValueSetController
         response.setContentType( CONTENT_TYPE_XML );
 
         DataExportParams params = dataValueSetService.getFromUrl( dataSet, dataElementGroup,
-            period, startDate, endDate, orgUnit, children, orgUnitGroup, includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
+            period, startDate, endDate, orgUnit, children, orgUnitGroup, attributeOptionCombo, 
+            includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
 
         dataValueSetService.writeDataValueSetXml( params, response.getOutputStream() );
     }
@@ -154,6 +156,7 @@ public class DataValueSetController
         @RequestParam( required = false ) Set<String> orgUnit,
         @RequestParam( required = false ) boolean children,
         @RequestParam( required = false ) Set<String> orgUnitGroup,
+        @RequestParam( required = false ) Set<String> attributeOptionCombo,
         @RequestParam( required = false ) boolean includeDeleted,
         @RequestParam( required = false ) Date lastUpdated,
         @RequestParam( required = false ) String lastUpdatedDuration,
@@ -163,7 +166,8 @@ public class DataValueSetController
         response.setContentType( CONTENT_TYPE_JSON );
 
         DataExportParams params = dataValueSetService.getFromUrl( dataSet, dataElementGroup,
-            period, startDate, endDate, orgUnit, children, orgUnitGroup, includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
+            period, startDate, endDate, orgUnit, children, orgUnitGroup, attributeOptionCombo,
+            includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
 
         dataValueSetService.writeDataValueSetJson( params, response.getOutputStream() );
     }
@@ -178,6 +182,7 @@ public class DataValueSetController
         @RequestParam( required = false ) Set<String> orgUnit,
         @RequestParam( required = false ) boolean children,
         @RequestParam( required = false ) Set<String> orgUnitGroup,
+        @RequestParam( required = false ) Set<String> attributeOptionCombo,
         @RequestParam( required = false ) boolean includeDeleted,
         @RequestParam( required = false ) Date lastUpdated,
         @RequestParam( required = false ) String lastUpdatedDuration,
@@ -188,7 +193,8 @@ public class DataValueSetController
         response.setContentType( CONTENT_TYPE_CSV );
 
         DataExportParams params = dataValueSetService.getFromUrl( dataSet, dataElementGroup,
-            period, startDate, endDate, orgUnit, children, orgUnitGroup, includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
+            period, startDate, endDate, orgUnit, children, orgUnitGroup, attributeOptionCombo,
+            includeDeleted, lastUpdated, lastUpdatedDuration, limit, idSchemes );
 
         dataValueSetService.writeDataValueSetCsv( params, response.getWriter() );
     }
diff --git a/dhis-2/dhis-web/dhis-web-importexport/src/main/java/org/hisp/dhis/importexport/action/datavalue/ExportDataValueAction.java b/dhis-2/dhis-web/dhis-web-importexport/src/main/java/org/hisp/dhis/importexport/action/datavalue/ExportDataValueAction.java
index dc4b01757..139991fa1 100644
--- a/dhis-2/dhis-web/dhis-web-importexport/src/main/java/org/hisp/dhis/importexport/action/datavalue/ExportDataValueAction.java
+++ b/dhis-2/dhis-web/dhis-web-importexport/src/main/java/org/hisp/dhis/importexport/action/datavalue/ExportDataValueAction.java
@@ -145,7 +145,7 @@ public class ExportDataValueAction
         HttpServletResponse response = ServletActionContext.getResponse();
 
         DataExportParams params = dataValueSetService.getFromUrl( selectedDataSets, null, null,
-            getMediumDate( startDate ), getMediumDate( endDate ), orgUnits, true, null, false, null, null, null, idSchemes );
+            getMediumDate( startDate ), getMediumDate( endDate ), orgUnits, true, null, null, false, null, null, null, idSchemes );
 
         if ( FORMAT_CSV.equals( exportFormat ) )
         {
